pkg/controller/schedulingmanager/controller.go

```diff
@@ -152,76 +182,56 @@ func (c *SchedulerController) reconcile(qualifiedName util.QualifiedName) util.R
 	if !ok {
-		var err error
-
-		scheduler, err = schedulingpreference.StartSchedulingPreferenceController(c.config, *schedulingType)
+		glog.Infof("Starting schedulingpreference controller for %s", schedulingKind)
+		stopChan := make(chan struct{})
+		schedulerInterface, err := schedulingpreference.StartSchedulingPreferenceController(c.config, *schedulingType, stopChan)
 		if err != nil {
-			runtime.HandleError(errors.Wrapf(err, "Error starting schedulingpreference controller for %q", schedulingKind))
+			runtime.HandleError(errors.Wrapf(err, "Error starting schedulingpreference controller for %s", schedulingKind))
 			return util.StatusError
 		}
-		c.scheduler[schedulingKind] = scheduler
+		abstractScheduler = newSchedulerWrapper(schedulerInterface, stopChan)
+		c.schedulers.Store(schedulingKind, abstractScheduler)
 	}
 
-	federatedKind := typeConfig.GetFederatedType().Kind
-	glog.Infof("Starting plugin kind %s for scheduling type %s", federatedKind, schedulingKind)
+	scheduler := abstractScheduler.(*SchedulerWrapper)
+	if scheduler.HasPlugin(typeConfigName) {
+		// Scheduler and plugin already running for this target typeConfig
+		return util.StatusAllOK
+	}
 
+	federatedKind := typeConfig.GetFederatedType().Kind
+	glog.Infof("Starting plugin %s for %s", federatedKind, schedulingKind)
 	err = scheduler.StartPlugin(typeConfig)
 	if err != nil {
-		runtime.HandleError(errors.Wrapf(err, "Error starting plugin for %q", federatedKind))
+		runtime.HandleError(errors.Wrapf(err, "Error starting plugin %s for %s", federatedKind, schedulingKind))
 		return util.StatusError
 	}
-	c.runningPlugins.Insert(qualifiedName.Name)
-	c.federatedKindMap[qualifiedName.Name] = federatedKind
+	scheduler.pluginMap.Store(typeConfigName, federatedKind)
 
 	return util.StatusAllOK
 }
 
-func (c *SchedulerController) stopScheduler(schedulingKind string, qualifiedName util.QualifiedName) {
-	c.Lock()
-	defer c.Unlock()
-	scheduler, ok := c.scheduler[schedulingKind]
+func (c *SchedulingManager) stopScheduler(schedulingKind, typeConfigName string) {
+	abstractScheduler, ok := c.schedulers.Get(schedulingKind)
 	if !ok {
 		return
 	}
 
-	if !c.runningPlugins.Has(qualifiedName.Name) {
-		return
-	}
-
-	kind, ok := c.federatedKindMap[qualifiedName.Name]
-	if ok {
-		glog.Infof("Stopping plugin for %q with kind %q", qualifiedName.Name, kind)
-		scheduler.StopPlugin(kind)
-		delete(c.federatedKindMap, qualifiedName.Name)
-	}
-	c.runningPlugins.Delete(qualifiedName.Name)
-
-	// if all resources registered to same scheduler are deleted, the scheduler should be stopped
-	resources := schedulingtypes.GetSchedulingKinds(qualifiedName.Name)
-	result := c.runningPlugins.Intersection(resources)
-	if result.Len() == 0 {
-		glog.Infof("Stopping scheduler schedulingpreference controller for %q", schedulingKind)
-		scheduler.Stop()
-
-		delete(c.scheduler, schedulingKind)
+	scheduler := abstractScheduler.(*SchedulerWrapper)
+	if scheduler.HasPlugin(typeConfigName) {
```

Comment:
(No action required) I think `HasPlugin` is redundant here:

```golang
kind, ok := scheduler.pluginMap.Get(typeConfigName)
if !ok {
  ...
}
```

(No action required) I also think kind should be cast once rather than in multiple places.